### PR TITLE
Implementing issue #10

### LINKED ISSUE #10: Modernización del repositorio: actualizaciones de GitHub Actions, Docker, dependencias y código
# Modernización del repositorio: actualizaciones de GitHub Actions, Docker, dependencias y código

  ## Contexto

  Este repositorio tiene aproximadamente 7 años desde su última actualización significativa. Varias de las herramientas y versiones que utiliza han llegado a su fin de vida (EOL) o han sido deprecadas. Este issue documenta un plan de mejora completo organizado por prioridad.

  ---

  ## Diagnóstico general

  El repositorio tiene deuda técnica en 4 áreas: **GitHub Actions**, **Docker**, **dependencias Python**, y **código/tests**.

  ---

  ## Área 1 — GitHub Actions

  | Problema | Archivo | Acción |
  |---|---|---|
  | `actions/checkout@v1` (2019) | ambos workflows | Actualizar a `actions/checkout@v4` |
  | `actions/setup-python@v1` | `run_tests.yml` | Actualizar a `actions/setup-python@v5` |
  | Runner `ubuntu-18.04` deprecado | `run_tests.yml` | Cambiar a `ubuntu-latest` |
  | Python 3.7 (EOL desde jun 2023) | `run_tests.yml` | Actualizar a 3.12 o matrix 3.11/3.12/3.13 |                                                                                          
  | Sin cache de dependencias pip | ambos | Agregar `cache: 'pip'` en `setup-python` |                                                                                                       
  | Sin matrix testing | `run_tests.yml` | Agregar `strategy.matrix` para múltiples versiones de Python |                                                                                    
  | Trigger `push` comentado | `run_tests.yml` | Habilitar tests también en push a `master` |                                                                                                
  | Sin linting en CI | ambos | Agregar paso con `ruff` |                                                                                                                                    
  | Sin reporte de cobertura | `run_tests.yml` | Agregar `pytest-cov` y publicar resultado |                                                                                                 
  | Tag Docker con timestamp `$(date +%s)` | `dockerimage.yml` | Reemplazar con `${{ github.sha }}` para trazabilidad |                                                                      
                                                                                                                                                                                             
  ---             
                                                                                                                                                                                             
  ## Área 2 — Docker

  | Problema | Archivo | Acción |                                                                                                                                                            
  |---|---|---|
  | `FROM ubuntu:18.04` (EOL abr 2023) | `some_docker.Dockerfile` | Cambiar a `python:3.12-slim` |                                                                                           
  | Múltiples `RUN apt-get` separados | `some_docker.Dockerfile` | Consolidar en una sola instrucción para reducir capas |                                                                   
  | Instalación manual de Python vía apt | `some_docker.Dockerfile` | Eliminar: `python:3.12-slim` ya incluye Python |
  | Sin `.dockerignore` | — | Crear `.dockerignore` para excluir `.git/`, `venv/`, `__pycache__/`, `tests/` |                                                                                
  | `PYTHONPATH` definido en `CMD` | `some_docker.Dockerfile` | Mover a `ENV PYTHONPATH=/code` |                                                                                             
  | Uso de `pip3` en lugar de `pip` | `some_docker.Dockerfile` | En imagen oficial `python:` usar simplemente `pip` |                                                                        
                                                                                                                                                                                             
  ---                                                                                                                                                                                        
                                                                                                                                                                                             
  ## Área 3 — Dependencias Python

  | Problema | Acción |
  |---|---|
  | `pytest==5.3.1` — versión actual: `9.0.3` | Actualizar a `pytest==9.0.3` |
  | Sin `pyproject.toml` | Crear `pyproject.toml` (PEP 517/518/621) como estándar moderno, centralizando configuración de pytest, ruff y mypy |                                              
  | Sin separación de dependencias dev/prod | Agregar sección `[project.optional-dependencies]` o `requirements-dev.txt` |                                                                   
  | Sin linter ni formatter | Agregar `ruff` (reemplaza flake8 + black + isort) |                                                                                                            
  | Sin type checking | Agregar `mypy` |                                                                                                                                                     
  | Sin medición de cobertura | Agregar `pytest-cov` |                                                                                                                                       
                                                                                                                                                                                             
  ---             
                                                                                                                                                                                             
  ## Área 4 — Código y tests

  | Problema | Ubicación | Acción |
  |---|---|---|
  | Typo `"Setieng value to 5"` | `some_script.py:13` | Corregir a `"Setting value to 5"` |
  | Tests acceden a `_some_att` (atributo privado) | `test_some_class.py:7,11,16` | Refactorizar para usar solo la API pública (`get_att()`) |                                               
  | Patrón getter/setter estilo Java | `some_class.py` | Refactorizar a `@property` (idioma Python estándar) |                                                                               
  | Sin `conftest.py` | — | Crear `conftest.py` para fixtures compartidos |                                                                                                                  
  | Sin configuración de pytest | — | Agregar sección `[tool.pytest.ini_options]` en `pyproject.toml` con `testpaths` y `pythonpath` |                                                       
                                                                                                                                                                                             
  ---                                                                                                                                                                                        
                                                                                                                                                                                             
  ## Plan de acción priorizado

  ### Prioridad alta — rompen CI o son EOL                                                                                                                                                   
  - [ ] Actualizar `actions/checkout` a v4 y `actions/setup-python` a v5
  - [ ] Cambiar runner `ubuntu-18.04` → `ubuntu-latest`                                                                                                                                      
  - [ ] Actualizar Python 3.7 → 3.12
  - [ ] Cambiar base Docker `ubuntu:18.04` → `python:3.12-slim`                                                                                                                              
  - [ ] Crear `.dockerignore`                                                                                                                                                                
  
  ### Prioridad media — calidad y modernización                                                                                                                                              
  - [ ] Actualizar `pytest==5.3.1` → `pytest==9.0.3`
  - [ ] Crear `pyproject.toml` con configuración centralizada                                                                                                                                
  - [ ] Agregar cache de pip en CI
  - [ ] Agregar matrix testing (Python 3.11, 3.12, 3.13)                                                                                                                                     
  - [ ] Mover `PYTHONPATH` de `CMD` a `ENV` en Dockerfile                                                                                                                                    
  - [ ] Consolidar capas `RUN` en Dockerfile                                                                                                                                                 
                                                                                                                                                                                             
  ### Prioridad baja — buenas prácticas                                                                                                                                                      
  - [ ] Agregar `ruff` como linter/formatter y paso en CI
  - [ ] Agregar `pytest-cov` y reporte de cobertura                                                                                                                                          
  - [ ] Refactorizar `SomeClass` para usar `@property`
  - [ ] Corregir tests para no acceder a `_some_att` directamente                                                                                                                            
  - [ ] Corregir typos en `some_script.py`                                                                                                                                                   
  - [ ] Usar `${{ github.sha }}` como tag Docker en lugar de timestamp

### PR DESCRIPTION
**Pull request by mperalta92**
# 🚀 Descripción       
                                                                                  
  Modernización completa del repositorio: actualización de GitHub Actions, Docker, dependencias Python y refactorización del código. El repo tenía ~7 años de deuda técnica con versiones EOL
   en todas las capas.                                                                                                                                                                       
                                                                                                                                                                                             
  Closes #10                                                                                                                                                                                   
                                                                                                                                                                                             
  ---                                                                                                                                                                                        
                                                            
  ### 🛠️  Tipo de Cambio

  - [ ] ✨ Nueva funcionalidad (Feature)
  - [ ] 🐛 Corrección de error (Bug fix)
  - [x] 🧹 Refactorización / Deuda técnica                                                                                                                                                   
  - [ ] 📝 Documentación
                                                                                                                                                                                             
  ---                                                       

  ## 🧪 Checklist de Pruebas                                                                                                                                                                 
  
  - [x] **Unit Tests:** 14/14 tests pasan localmente con pytest 9.0.3 y Python 3.14                                                                                                          
  - [x] **Manual QA:** Pipeline de CI verificado en estructura; Docker build validado con nueva imagen base
  - [x] **Edge Cases:** Tests cubren valores negativos, instancias independientes, idempotencia y sobrescritura de valores                                                                   
                                                                                                                                                                                             
  ---                                                                                                                                                                                        
                                                                                                                                                                                             
  ## 📸 Evidencia (Screenshots/Logs)                        

  ============================= test session starts ==============================
  platform darwin -- Python 3.14.3, pytest-9.0.3, pluggy-1.6.0
  configfile: pyproject.toml                                                                                                                                                                 
  testpaths: tests
  collected 14 items                                                                                                                                                                         
                                                            
  tests/some_module/test_some_class.py::test_default_initial_value PASSED                                                                                                                    
  tests/some_module/test_some_class.py::test_custom_initial_value PASSED
  tests/some_module/test_some_class.py::test_property_setter PASSED                                                                                                                          
  tests/some_module/test_some_class.py::test_property_setter_overwrites PASSED
  tests/some_module/test_some_class.py::test_increment PASSED                                                                                                                                
  tests/some_module/test_some_class.py::test_multiple_increments PASSED
  tests/some_module/test_some_class.py::test_negative_initial_value PASSED                                                                                                                   
  tests/some_module/test_some_class.py::test_increment_from_negative PASSED                                                                                                                  
  tests/some_module/test_some_class.py::test_instances_are_independent PASSED
  tests/some_module/test_some_class.py::test_fixture_valued_instance PASSED                                                                                                                  
  tests/some_module/test_some_class.py::test_set_then_increment PASSED
  tests/test_some_script.py::test_run_returns_expected_value PASSED                                                                                                                          
  tests/test_some_script.py::test_run_returns_int PASSED                                                                                                                                     
  tests/test_some_script.py::test_run_is_idempotent PASSED
                                                                                                                                                                                             
  Name                                  Stmts   Miss  Cover   Missing                                                                                                                        
  
  some_code/init.py                     0      0   100%                                                                                                                                      
  some_code/some_module/init.py         1      0   100%     
  some_code/some_module/some_class.py      11      0   100%
  some_code/some_script.py                 13      1    92%   20                                                                                                                             
  
  TOTAL                                    25      1    96%                                                                                                                                  
  ============================== 14 passed in 0.02s ==============================
                                                                                                                                                                                             
  ---
                                                                                                                                                                                             
  ## ⚠️  Notas para el Revisor                               

  ### Cambios con impacto en uso del repo                                                                                                                                                    
  
  - **`SomeClass` API:** Los métodos `get_att()` y `set_att()` fueron reemplazados por un `@property att`. Cualquier código externo que use la clase deberá actualizar las llamadas:         
  `obj.set_att(5)` → `obj.att = 5`, `obj.get_att()` → `obj.att`.
  - **`requirements.txt`:** Ya no contiene dependencias. Todo se gestiona desde `pyproject.toml`. Instalar con `pip install -e ".[dev]"`.                                                    
  - **Docker:** La imagen base cambió de `ubuntu:18.04` a `python:3.12-slim`. El build es más rápido y liviano al eliminar la instalación manual de Python vía apt.                          
  - **GitHub Actions:** El workflow `Run Tests` ahora corre en matrix (Python 3.11, 3.12, 3.13) y también se dispara en `push` a `master`, no solo en PRs.                                   
                                                                                                                                                                                             
  ### Sin impacto en infraestructura                                                                                                                                                         
  - **Migraciones:** No                                                                                                                                                                      
  - **Variables de entorno:** No hay variables de entorno nuevas requeridas

### CODE REVIEW

File: .github/workflows/run_tests.yml [hunk 1]
```diff
@@ -1,26 +1,39 @@
...
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+          cache: "pip"
+
+      - name: Install dependencies
+        run: pip install -e ".[dev]"
+
+      - name: Lint with ruff
+        run: ruff check .
```
Comment (FBernuy) [inline .github/workflows/run_tests.yml:31]: 🤟🏽

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+[tool.ruff]
+target-version = "py311"
+line-length = 88
+
+[tool.ruff.lint]
+select = ["E", "F", "I", "UP"]
+
+[tool.mypy]
+python_version = "3.12"
```
Comment (mperalta92) [inline pyproject.toml]:
```suggestion
python_version = "3.11"
```

File: .github/workflows/run_tests.yml [hunk 3]
```diff
@@ -1,26 +1,39 @@
...
-    - name: Test with pytest
-      run: |
-        python -m pytest
+      - uses: actions/checkout@v4
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+          cache: "pip"
```
Comment (mperalta92) [inline .github/workflows/run_tests.yml:24]:
```suggestion
          cache: "pip"
          cache-dependency-path: pyproject.toml
```

**Comment by mperalta92** [comment]
Aplicados los cambios sugeridos!! , @FBernuy  dele merge cuando estime conveniente.
Un abrazo
